src/KUSignalLib/DSP/DSP.py

```diff
@@ -25,4 +26,114 @@ def IIRDirectForm2(b, a, x):
+    :param mode: Interpolation type -> "linear", "quadratic"
+    :return: Interpolated signal.
+    """
+    nonzero_indices = np.arange(0, len(x)*n, n) # Generate indices for upsampled signal
+    interpolation_function = intp.interp1d(nonzero_indices, np.array(x), kind=mode, fill_value='extrapolate') # create interpolation function
+    interpolated_signal = interpolation_function(np.arange(len(x)*n)) # interpolate the signal
+    return interpolated_signal
+
+
+def Upsample(x, n, dim=1, interpolate=True, axis=0):
+    """
+    Discrete signal upsample implementation.
+
+    :param x: Input signal.
+    :param n: Upsample factor.
+    :param dim: Dimension of input signal. Default is 1.
+    :param interpolate: Flag indicating whether to perform interpolation. Default is True.
+    :param axis: Axis of x that is upsampled. Default is 0.
+    :return: Upsampled signal.
+    """
+    x_upsampled = []  # Initialize a list to store the upsampled signal
+
+    if dim == 1:  # If the input signal is 1D
+        if interpolate:
+            x_upsampled = Interpolate(x, n, mode="linear")
+        else:
+            for i in range(len(x)):  # Iterate over each element in the input signal
+                x_upsampled += [x[i]] + list(np.zeros(n-1, dtype=type(x[0])))  # Add the current element and n zeros after each element
+        return x_upsampled
+
+    elif dim == 2:  # If the input signal is 2D
+        if axis == 0:  # If axis input is 0, interpolate over rows
+            if interpolate:
+                for i in range(len(x)):
+                    x_upsampled += [list(Interpolate(x[i], n, mode="linear"))]
+            else:
+                for i in range(len(x)):  # Iterate over each row in the input signal
+                    x_i_upsampled = []  # Initialize a list to store the upsampled row
+                    for j in range(len(x[i])):  # Iterate over each element in the current row
+                        x_i_upsampled += [x[i][j]] + list(np.zeros(n-1, dtype=type(x[i][0]))) # Add the current element and n zeros after each element
+                    x_upsampled.append(x_i_upsampled)  # Add the upsampled row to the upsampled signal
+            return x_upsampled
+
+        elif axis == 1:  # If axis input is 1, interpolate over columns
+            if interpolate:
+                pass
+            else:
+                for i in range(len(x)):
+                    for j in range(n-1):
+                        x_upsampled += [x[i]]
+                        for k in range(n-1):
+                            x_upsampled += [list(np.zeros(len(x[i]), dtype=type(x[i][0])))]
+            return x_upsampled
+
+        else:
+            raise ValueError("Invalid axis input.")  # Raise an error indicating invalid axis input
+    else:  # If the input dimension is not 1 or 2
+        raise ValueError("Invalid dimension input.")  # Raise an error indicating invalid dimension input
+
+
+def Downsample(x, n, dim=1, axis=0):
```

Comment:
alot of this is gonna be a repeat from above, and i dont want to write it all, but sample offset, auto-detect input size, book uses M for down sample factor